[andrefoo] iP

andrefoo opened the pull request:
# Dre 

"Your mind is for having ideas, not holding them.” — David Allen
Dre frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/andrefoo/ip/releases/download/A-Release/dre.zip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

andrefoo committed:
Level-0
Rename, Greet, Exit

andrefoo committed:
Level-1
Echo

andrefoo committed:
Level-2
Add, List

andrefoo committed:
Level-3 and 4
Mark as Done
ToDos, Events, Deadlines

andrefoo committed:
A-TextUiTesting: Automated Text Ui Testing

andrefoo committed:
Level 5: Error Handling

andrefoo committed:
Level 6: Support Delete feature

andrefoo committed:
Update Dre.java

andrefoo committed:
Level 7: Save

andrefoo committed:
Level 7: Save

andrefoo committed:
Add date feature

Deadline and Event date stored in LocalDate format

andrefoo committed:
Delete comments

andrefoo committed:
Add more OOP

Improved OOP in a more user oriented fashion.
Add Command classes
Add Ui class
Add Storage class
Add Parser class
Add TaskList class

jinyuan0425 commented on src/main/java/MarkCommand.java:
```diff
@@ -0,0 +1,17 @@
+public class MarkCommand extends Command {
+    private final int index;
```
Perhaps the final keyword can be omitted?

jinyuan0425 commented on src/main/java/UnmarkCommand.java:
```diff
@@ -0,0 +1,17 @@
+public class UnmarkCommand extends Command {
+    private final int index;
```
Perhaps the final keyword can be omitted?

jinyuan0425 commented on src/main/java/Dre.java:
```diff
@@ -0,0 +1,242 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Dre {
```
You may consider deleting unused files to minimize confusion.

jinyuan0425 reviewed (commented):
To improve readability, perhaps you can include headers for documentation of you code.

yarnmengnus commented on src/main/java/Dre.java:
```diff
@@ -0,0 +1,242 @@
+                    }
+                } catch (IOException e) {
+                    System.out.println("Error reading tasks from file.");
+                }
```
try to reduce nesting here

yarnmengnus commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Task {
+    private String task;
+    private boolean done;
```
could change to isDone

yarnmengnus commented on src/main/java/Dre.java:
```diff
@@ -0,0 +1,242 @@
+                writer.newLine();
+            }
+        } catch (IOException e) {
+            System.out.println("Error saving tasks to the file.");
```
Could abstract to ui

yarnmengnus commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,80 @@
+    public static Task parseTask(String line) {
+        String type = line.substring(0, 1);
+        boolean isDone = line.charAt(2) == 'X';
+        String description;
```
Could initialise to empty string

andrefoo committed:
Organize classes into dedicated packages

Refactored the codebase to introduce a modular package structure.

- Created 'dre' for Dre main interface and subpackages.
- Created `dre.task` for task-related classes.
- Created `dre.command` for command-related classes.
- Created 'dre.parser' for the parsing classes.
- Created 'dre.exception' for chatbot specific exceptions.
- Created 'dre.ui' for ui related classes.

This structure will help in better organizing the project as it scales and provides clearer segregation of responsibilities.

andrefoo committed:
Merge branch 'add-gradle-support'

andrefoo committed:
Amend build.gradle file access name

Clin-lyx commented on src/main/java/dre/Dre.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Dre {
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
Perhaps it would be better if you made these three variables final. (Remember to CAP all letters :) )

Clin-lyx commented on src/main/java/dre/command/MarkCommand.java:
```diff
@@ -0,0 +1,24 @@
+import dre.exception.DreException;
+
+public class MarkCommand extends Command {
+    private final int index;
```
Please CAP final variables

Clin-lyx commented on src/main/java/dre/command/DeleteCommand.java:
```diff
@@ -0,0 +1,24 @@
+import dre.task.TaskList;
+
+public class DeleteCommand extends Command {
+    private final int index;
```
Please CAP index here :)

Clin-lyx commented on src/main/java/dre/task/Task.java:
```diff
@@ -0,0 +1,43 @@
+
+public class Task {
+    private String task;
+    private boolean done;
```
Please change it to isDone to fit the coding standard.

Clin-lyx commented on src/main/java/dre/task/Task.java:
```diff
@@ -0,0 +1,43 @@
+        done = false;
+    }
+
+    public boolean isDone() {
```
Should this method be deleted? I think it is not used anywhere, sorry if I am wrong

Clin-lyx commented on src/main/java/dre/command/Command.java:
I like how you design your command class 👍

Clin-lyx reviewed (commented):
Overall, I think you did a good job implementing these classes. Only some minor violations of the coding standard. Please keep up the good work!!!

andrefoo committed:
Add unit tests for Task package classes

This commit introduces unit tests for the classes within the Task package:

- `TaskTest`: Tests basic task functionalities like marking as done.
- `DeadlineTest`: Tests specific deadline functionalities, e.g., date retrieval.
- `EventTest`: Tests specific event functionalities, like event date checking.
- `ToDoTest`: Tests todo-specific behaviors, like description retrieval.

These tests aim to ensure that the basic methods and behaviors of task-related classes function as expected and help in identifying potential regressions in future refactoring or feature additions.

andrefoo committed:
Amend tests for dre.test package

testToString bug: changed printed date format
testFileSaveFormat bug: LocalDate.parse input date format changed to yyyy-MM-dd

andrefoo committed:
Delete unused imports

Parser: java.util.Arrays not utilised.
DeadlineTest: junit BeforeAll not utilised.
ToDoTest: java.time.LocalDate not utilised.

andrefoo committed:
Include final keyword on variables

Added final keyword for some unchanging variables in Dre, Deadline, Event.

andrefoo committed:
Add JavaDoc comments for all classes

andrefoo committed:
Implement Java Coding Standard

Shortened lines that were near 120 characters
Removed wildcard imports

andrefoo committed:
Add find function in chatbot

Add FindCommand
Edit Parser.parse to include find command
Add a few methods in Task and TaskList to accommodate word finding
Add showFoundTasks method in Ui

This helps the user easily find tasks in a longer list.

andrefoo committed:
Resolve merge conflict between A-CodingStandard and Level-9

andrefoo committed:
Resolve merge conflict between master and A-JavaDocs

andrefoo committed:
Add import in Parser

Parser: cannot resolve symbol 'FindCommand'

Cannot compile. Add import for FindCommand to resolve error.

andrefoo committed:
Update JavaDocs for Level-9 functions

Add Javadocs comments for FindCommand.

andrefoo committed:
Implement JavaFX for new GUI

Adds JavaFX compatibility

andrefoo committed:
Incorporate Java assertions

Certain application invariants in the Dre and Parser classes are not validated.

Unchecked assumptions can lead to subtle bugs.

Let's add Java assertions to:
- Validate the initialization of storage, tasks, and image resources in `Dre.java`.
- Ensure input strings in `Parser.java` are valid and parsed objects are non-null.

Assertions provide early feedback during development, ensuring application invariants hold true, making the system more robust.

andrefoo committed:
Refactor multiple classes for clarity and maintainability

Several classes in the project contained magic numbers, long lines, and lacked encapsulation.

Let's:
    Replace magic numbers with named constants for better clarity.
    Break up long lines to enhance code readability.
    Encapsulate specific functionalities into private methods.

Doing so enhances code readability, reduces potential errors, and makes the codebase easier to maintain.

andrefoo committed:
Merge pull request #3 from andrefoo/branch-A-CodeQuality

Merge branch-A-CodeQuality

andrefoo committed:
Merge pull request #2 from andrefoo/branch-A-Assertions

Merge branch-A-Assertions

andrefoo committed:
Update code quality

Some classes lack comprehensive
documentation and do not follow naming
conventions.

To improve code readability and support
developers:
- Let's add Javadoc comments to these classes,
providing clear explanations and details on
their functionalities.
- refactor variable names such as 'done' to
'isDone' to explain themselves

andrefoo committed:
Merge pull request #4 from andrefoo/iP-code-quality-feedback

Update code quality

andrefoo committed:
Implement task editing functionality

Tasks can only be added or removed.
To enhance user flexibility in managing tasks,
let's introduce the ability to edit task details:

    Enable description editing for all tasks.
    Allow date modifications for Deadline and
Event tasks.

This approach provides users with more
control over their tasks without needing to
delete and re-add tasks for minor changes.

andrefoo committed:
Merge pull request #5 from andrefoo/branch-BCD-extension

Implement task editing functionality

andrefoo committed:
Update readability

Since updating to a JavaFX GUI, some methods do not need to be used.

Let's remove unused methods in Dre.java and update JavaDocs for edited
tasks.

andrefoo committed:
Update readability and fix bugs

Unused methods in a few classes.

Since updating to a JavaFX GUI, some methods do not need to be used.

Let's remove unused methods in ExitCommand.java and update missing
JavaDocs in newer classes like EditCommand.java and newer methods like
editByDate() in Deadline.java.

Edit command: date edits not working

Edit command for dates are not working.

Let's refactor code in EditCommand to read the correct cases and add a
default case that throws error if the field is not recognised.

andrefoo committed:
Add User Guide for Dre Chatbot

The project lacks user-facing documentation.

To improve user experience and usability,

Let's add a user guide that explains features and their usage, and also
provides example commands and expected outcomes.

This basic guide will help both new and existing users understand how to
interact with the chatbot effectively.

andrefoo committed:
Handle data storage exceptions

Storage.java: Throws error if ./data/dre.txt not found.

Make Storage.java create the file itself if it doesn't exist.

Let's add methods to check if the storage file exists, and create the
file if it doesn't.

This allows the application to exit normally and the jar file now
operates independently of any other folder.

andrefoo committed:
Update README.md

Add acknowledgements
